src/full/Agda/TypeChecking/Sort.hs

```diff
@@ -85,57 +86,66 @@ hasBiggerSort = void . inferUnivSort
 --   return that. Otherwise, return @PiSort a s2@ and add a constraint to
 --   ensure we can compute the sort eventually.
-inferPiSort :: PureTCM m => Dom Type -> Abs Sort -> m Sort
+inferPiSort :: (PureTCM m,MonadConstraint m) => Dom Type -> Abs Sort -> m Sort
 inferPiSort a s2 = do
-  s1' <- reduce $ getSort a
-  s2' <- mapAbstraction a reduce s2
-  -- we do instantiateFull here to perhaps remove some (flexible)
-  -- dependencies of s2 on var 0, thus allowing piSort' to reduce
-  s2' <- instantiateFull s2'
-
+  s1' <- reduceB $ getSort a
+  s2' <- mapAbstraction a reduceB s2
   --Jesper, 2018-04-23: disabled PTS constraints for now,
   --this assumes that piSort can only be blocked by unsolved metas.
+  --Arthur Adjedj, 2023-02-27, Turned PTS back on,
+  --piSort can now be blocked by Leveluniv
+  let b' = unblockOnEither (getBlocker s1') (getBlocker $ unAbs s2')
+  case piSort' (unEl <$> a) (ignoreBlocking s1') (ignoreBlocking <$> s2') of
+    Right s -> return s
+    Left b -> do
+      addConstraint (unblockOnEither b b') $ HasPTSRule a (ignoreBlocking <$> s2')
+      return $ PiSort (unEl <$> a) (ignoreBlocking s1') (ignoreBlocking <$> s2')
 
-  --case piSort' s1 s2 of
-  --  Just s -> return s
-  --  Nothing -> do
-  --    addConstraint $ HasPTSRule s1 s2
-  --    return $ PiSort s1 s2
-
-  return $ piSort (unEl <$> a) s1' s2'
 
 -- | As @inferPiSort@, but for a nondependent function type.
 inferFunSort :: Sort -> Sort -> TCM Sort
-inferFunSort s1 s2 = funSort <$> reduce s1 <*> reduce s2
-
-ptsRule :: Dom Type -> Abs Sort -> Sort -> TCM ()
-ptsRule a b c = do
-  c' <- inferPiSort a b
-  ifM (optCumulativity <$> pragmaOptions)
-    (leqSort c' c)
-    (equalSort c' c)
-
--- | Non-dependent version of ptsRule
-ptsRule' :: Sort -> Sort -> Sort -> TCM ()
-ptsRule' a b c = do
-  c' <- inferFunSort a b
-  ifM (optCumulativity <$> pragmaOptions)
-    (leqSort c' c)
-    (equalSort c' c)
+inferFunSort s1 s2 = do
+  ss1 <- sortOf (Sort s1)
+  s1' <- reduceB s1
+  s2' <- reduceB s2
+  case funSort' (ignoreBlocking s1') (ignoreBlocking s2') of
+    Right s -> return s
+    Left b -> do
+      let b' = unblockOnEither (getBlocker s1') (getBlocker s2')
+      addConstraint (unblockOnEither b b') $ HasPTSRule (defaultDom (El (univSort s1) (Sort s1))) (mkAbs "_" $ ignoreBlocking s2')
+      return $ FunSort s1 s2
 
 hasPTSRule :: Dom Type -> Abs Sort -> TCM ()
-hasPTSRule a b = void $ inferPiSort a b
+hasPTSRule a b =
+  if alwaysValidCodomain $ unAbs b
+  then return ()
+  else do
+    infer <- reduceB =<< inferPiSort a b
+    case infer of
+      Blocked b t | neverUnblock == b -> typeError $ InvalidTypeSort t
+      NotBlocked _ t@FunSort{}        -> typeError $ InvalidTypeSort t
+      NotBlocked _ t@PiSort{}         -> typeError $ InvalidTypeSort t
+      _ -> return ()
+  where
+    alwaysValidCodomain s | Right LargeSort{} <- sizeOfSort s
+                               = True
```

Comment:
It would appear to me that `Inf` is the only large sort, so one could match on it directly...